paasta_tools/frameworks/constraints.py

```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env python
+# Without this, the import of mesos.interface breaks because paasta_tools.mesos exists
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import re
+
+from paasta_tools.utils import paasta_print
+
+
+def max_per(args):
+    cv, ov, at, st = args
+    if not cv:
+        cv = 1
+    return st.get('MAX_PER', {}).get(at, {}).get(ov, 0) <= int(cv)
+
+
+# lambda arg: [constraint value, offer value, attribute, state]
+CONS_OPS = {
+    'EQUALS': lambda pair: pair[0] == pair[1],
+    'LIKE': lambda pair: re.match(pair[0], pair[1]),
+    'UNLIKE': lambda pair: not(re.match(pair[0], pair[1])),
+
+    # example: ['pool', 'MAX_PER', 5]
+    #   constraint value: 5
+    #   offer value:      default
+    #   attribute:        pool
+    #   state:            {'MAX_PER' => {'pool' => {'default' => 6}}}
+    'MAX_PER': max_per,
+    'UNIQUE': max_per,
+}
+
+
+def nested_inc(op, args):
+    """Increments relevant counter by step from args array"""
+    _, attr_val, attr_name, state, step = args
+    oph = state.setdefault(op, {})
+    nameh = oph.setdefault(attr_name, {})
+    nameh.setdefault(attr_val, 0)
+    nameh[attr_val] += step
+    return state
+
+
+# lambda args same as CONS_OPS + update step
+UPDATE_OPS = {
+    'EQUALS': lambda _: None,
+    'LIKE': lambda _: None,
+    'UNLIKE': lambda _: None,
+    'MAX_PER': lambda x: nested_inc('MAX_PER', x),
+    'UNIQUE': lambda x: nested_inc('MAX_PER', x),
+}
+
+
+def test_offer_constraints(offer, constraints, state):
+    """Returns True if all constraints are satisfied by offer's attributes,
+    returns False otherwise. Prints a error message and re-raises if an error
+    was thrown."""
+    for (attr, op, val) in constraints:
+        try:
+            offer_attr = next(
+                (x for x in offer.attributes if x.name == attr), None)
+            if offer_attr is None:
+                paasta_print("Attribute not found for a constraint: %s" % attr)
+                return False
+            elif not(CONS_OPS[op]([val, offer_attr.text.value,
```

Comment:
argh, /me dumb. i thought lambda only supports single argument, probably because it only supports single statement. thanks